File: src/output/pr-draft.js
export function generatePrDraft(state) {
  const changedFiles = state.proposal?.changedFiles?.length
    ? state.proposal.changedFiles
    : extractFilesFromDiff(state.proposal?.workingDiff ?? state.proposal?.diff ?? "");
  const tests = state.testResult?.results?.length
    ? state.testResult.results.map((result) => `${result.ok ? "PASS" : "FAIL"} ${result.command}`)
    : ["Not run"];

  const title = state.issue.title.toLowerCase().startsWith("fix")
    ? state.issue.title
    : `fix: ${state.issue.title}`;

  const body = [
    "## Summary",
    "",
    `- ${state.intent.expectedBehaviour}`,
    `- Scoped through the repository graph to ${changedFiles.length ? changedFiles.join(", ") : "the planned files"}.`,
    "",
    "## Changes",
    "",
    ...(changedFiles.length ? changedFiles.map((file) => `- Updated \`${file}\`.`) : ["- No diff was applied in this run."]),
    "",
    "## Validation",
    "",
    ...tests.map((line) => `- ${line}`),
    "",
    `Fixes #${state.issue.number}`
  ].join("\n");

  return {
    title,
    body
  };
}

function extractFilesFromDiff(diff) {
  const files = new Set();
  for (const match of diff.matchAll(/^\+\+\+\s+b\/(.+)$/gm)) {
    files.add(match[1]);
  }
  return [...files];
}

File: go_issue_agent/output/pr_draft.py
import re


def generate_pr_draft(state: dict) -> dict:
    proposal = state.get("proposal") or {}
    changed_files = proposal.get("changedFiles") or _extract_files_from_diff(proposal.get("workingDiff") or proposal.get("diff", ""))
    if (state.get("testResult") or {}).get("results"):
        tests = [
            f'{"PASS" if result.get("ok") else "FAIL"} {result.get("command")}'
            for result in state["testResult"]["results"]
        ]
    else:
        tests = ["Not run"]

    title = state["issue"]["title"]
    if not title.lower().startswith("fix"):
        title = f"fix: {title}"

    body_lines = [
        "## Summary",
        "",
        f'- {state["intent"]["expectedBehaviour"]}',
        f'- Scoped through the repository graph to {", ".join(changed_files) if changed_files else "the planned files"}.',
        "",
        "## Changes",
        "",
    ]
    if changed_files:
        body_lines.extend([f"- Updated `{file}`." for file in changed_files])
    else:
        body_lines.append("- No diff was applied in this run.")
    body_lines.extend(["", "## Validation", ""])
    body_lines.extend([f"- {line}" for line in tests])
    body_lines.extend(["", f'Fixes #{state["issue"]["number"]}'])
    return {"title": title, "body": "\n".join(body_lines)}


def _extract_files_from_diff(diff: str) -> list[str]:
    return sorted(set(re.findall(r"^\+\+\+\s+b/(.+)$", diff, re.MULTILINE)))
